tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,365 @@
+# Copyright 2018 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Replica Exchange Monte Carlo Transition Kernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+import tensorflow as tf
+
+from tensorflow_probability.python.mcmc import kernel as kernel_base
+from tensorflow_probability.python.mcmc import util as mcmc_util
+from tensorflow.python.ops.distributions import util as distributions_util
+
+
+__all__ = [
+    'ReplicaExchangeMC',
+    'default_exchange_proposed_fn',
+]
+
+
+ReplicaExchangeMCKernelResults = collections.namedtuple(
+    'ReplicaExchangeMCKernelResults',
+    [
+        'replica_results',
+        'replica_states',
+        'next_replica_idx',
+        'exchange_proposed',
+        'exchange_proposed_n',
+    ])
+
+
+def default_exchange_proposed_fn(freq):
+  """Default function for `exchange_proposed_fn` of `kernel`.
+  Depending on the probability of `freq`, decide whether to propose combinations
+  of replica for exchange.
+  When exchanging, create combinations of adjacent replicas from 0 or 1 index.
+  """
+  def default_exchange_proposed_fn_(n_replica, seed=None):
+    seed = distributions_util.gen_new_seed(seed, "default_exchange_proposed_fn")
+    u = tf.random_uniform([], seed=seed)
+    exist = u > freq
```

Comment:
If `freq` is a non-scalar, I worry that this will do the wrong thing. I think we want to use:

    u = tf.random_uniform(tf.shape(freq), dtype=freq.dtype.base_dtype, seed=seed)

Reply:
Thank you for your comment!

I want to limit `freq` to scalar. Because exchanging is done for the states (set of variables) collectively rather than the individual variables.

I would appreciate it if you could reply.